lib/rspec/matchers/built_in/has.rb

```diff
@@ -46,8 +46,22 @@ def description
 
       private
 
+        # Catch a semi-frequent typo - if you have strict_predicate_matchers disabled and
+        # expect(spy).to have_receieveddd(:foo) it would be evergreen - the dynamic matcher
+        # queries `has_receiveddd?`, the spy _fakes_ the method, returning its (truthy) self.
+        def really_responds_to?(method)
+          if defined?(RSpec::Mocks::Double) && @actual.is_a?(RSpec::Mocks::Double)
+            @actual.respond_to?(method) && methods_include?(method)
+          else
+            @actual.respond_to?(method)
+          end
+        rescue NoMethodError
+          # Proxied BasicObjects don't have `is_a?`, or basically anything else.
+          @actual.respond_to?(method)
+        end
```

Comment:
We typically define conditional behaviour from existence / feature support by defining the method conditionally, also we should probably use the equality matcher here in case `is_a?` is tampered with, which would probably solve the basic object problem too.
```suggestion
        if defined?(RSpec::Mocks::Double)
          def really_responds_to?(method)
            if RSpec::Mocks::Doubl === @actual
              @actual.respond_to?(method) && methods_include?(method)
            else
              @actual.respond_to?(method)
            end
          rescue NoMethodError
            # Proxied BasicObjects don't have `is_a?`, or basically anything else.
            @actual.respond_to?(method)
          end
        else
          def really_responds_to?(method)
            @actual.respond_to?(method)
          end
        end
```

Reply:
Huh, BasicObject doesn't even have `respond_to?`, which makes that question kind of moot. `===` still works, so your solution is much more elegant, thanks :-)